Unify swarm init and update options

tonistiigi opened the pull request:
- Unify swarm init and update options
- Add api side validation and defaults for init and
  join requests.

@dnephin @aaronlehmann 

Signed-off-by: Tonis Tiigi tonistiigi@gmail.com

tonistiigi committed:
Unify swarm init and update options

Add api side validation and defaults for init and
join requests.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -1030,14 +1061,91 @@ func (c *Cluster) managerStats() (current bool, reachable int, unreachable int,
 	return
 }
 
+func validateInitRequest(req *types.InitRequest) error {
```
nit: This does mire than just validate thr request. It may also change it.

`validateAndSanitizeInitRequest`?

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -1030,14 +1061,91 @@ func (c *Cluster) managerStats() (current bool, reachable int, unreachable int,
+	}
+	if !strings.Contains(addr, ":") {
+		addr += ":" + strconv.Itoa(defaultPort)
+	}
```
This won't work with IPv6. We should use SplitHostPort.

aaronlehmann replied:
(and JoinHostPort)

aaronlehmann commented on integration-cli/docker_api_swarm_test.go:
```diff
@@ -656,6 +656,24 @@ loop0:
+	}
+	status, _, err = d.SockRequest("POST", "/swarm/join", req2)
+	c.Assert(err, checker.IsNil)
+	c.Assert(status, checker.Equals, http.StatusInternalServerError)
```
Let's add IPv6 test cases if it's minimal effort.

tonistiigi commented:
@aaronlehmann Updated. Needed to change the cli validation as well to make the ipv6 addresses work.

aaronlehmann commented:
@tonistiigi unit tests are failing.

tonistiigi committed:
Fix opts tests after default port fix

The code for default port was already there but
it didn’t work because split function errored out
before. This should be the desired behavior that
matches daemon listen address with swarm listen
address.

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi commented:
Added a commit with the `/opts` tests fixes. The code for the default port was already there but it didn't work because of a bug and testcases set all hosts without port as invalid. This should be the desired behavior that matches daemon listen address with swarm listen address.

aaronlehmann commented:
LGTM

AkihiroSuda commented on api/client/swarm/opts.go:
```diff
@@ -148,3 +141,24 @@ func (o *AutoAcceptOption) Policies(secret *string) []swarm.Policy {
+	flags.DurationVar(&opts.nodeCertExpiry, flagCertExpiry, time.Duration(90*24*time.Hour), "Validity period for node certificates")
+}
+
+func (opts *swarmOptions) ToSpec() swarm.Spec {
```
`Spec()` is more go-ish?
https://golang.org/doc/effective_go.html#interface-names

tonistiigi replied:
This is to match `service/opts.go` and `node/opts.go`. I don't want refactor those in this pr.

dnephin commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -74,3 +74,63 @@ func (s *DockerSwarmSuite) TestSwarmUpdate(c *check.C) {
+	out, err = d2.Cmd("info")
+	c.Assert(err, checker.IsNil, check.Commentf("out: %v", out))
+	c.Assert(out, checker.Contains, "Swarm: active")
+}
```
I think these could be unit tests instead of integration tests,n which would run a lot  faster.

vdemeester replied:
definitely 😇

tonistiigi replied:
I added a new unit test also for the ipv6 case. But this actually binds and connects as well so we can check that nothing goes wrong it the API levels.

LK4D4 commented:
well-known hang on janky

LK4D4 commented:
LGTM

thaJeztah commented:
@tonistiigi I take it this is for 1.12? Let me know if this wasn't intended for that
